packages/data/src/components/use-dispatch/use-dispatch-with-map.js

```diff
@@ -26,7 +32,9 @@ const useDispatchWithMap = ( dispatchMap ) => {
 						`Property ${ propName } returned from dispatchMap in useDispatchWithMap must be a function.`
 					);
 				}
-				return dispatcher;
+				return ( ...args ) => {
+					return currentDispatchMap.current( registry.dispatch, registry )[ propName ]( ...args );
```

Comment:
I wonder if there still is a flaw with this though.  `dispatchMap` might be the same function, but return a different collection of props between rerenders. In which case, this will only invoke the props functions returned on the intiial render (unless the function changes).

Eg. a map that looks something like this:
```js

const mapDispatch = ( dispatch ) => {
   if ( ! props.foo ) {
       return {};
   }
   return {
        onClick: dispatch.('someStore').addFoo( props.foo )
   };
};
```

Reply:
`dispatchMap` is a dependency there so it will re-run and return a new object.

What we can do is accept a `deps` argument and memoize `dispatchMap` at the top just like in `useSelect`.

Reply:
That would give it all the properties we want:

- The returned object with action dispatchers only changes when `dispatchMap` changes.
- `dispatchMap` only changes if it has `deps` and any of them change.

So for your example you would call it like this: `useDispatchWithMap( mapDispatch, [ props.foo ] )`.

Reply:
You also don't need any refs this way.

Reply:
Keep in mind:

- We're not exposing `useDispatchWithMap` for the general api, it's only intended for `withDispatch` internals.
- `withDispatch` can't pass any dependencies for the same reason it wouldn't work in the `withSelect` implementation of `useSelect` because props shape changes.
- My example is admittedly not clear because it would actually be something like this fed to `withDispatch` (and passed through here via `withDispatch`).

```js
const SomeWrappedComponent = withDispatch( ( dispatch, ownProps ) => {
    if ( ! ownProps.foo ) {
       return {};
    }
    return {
        onClick: dispatch.('someStore').addFoo( props.foo )
    };
} )( SomeComponent );
```

Reply:
For reference, here's the code that was failing in e2e tests which the `useRef` implementation fixed:

https://github.com/WordPress/gutenberg/blob/13e58516cc34dfd3fabc569ef2177c332091964e/packages/edit-post/src/components/keyboard-shortcut-help-modal/index.js#L101-L115

Reply:
In that case, this should happen:

- Prop changes make the wrapped component rerender.
- `mapDispatch` is redefined: https://github.com/WordPress/gutenberg/blob/de2d19d0713191b2887814594c56a4f169e4538d/packages/data/src/components/with-dispatch/with-dispatch-new.js#L14
- The hook's `useMemo` clears and returns a new object: https://github.com/WordPress/gutenberg/blob/de2d19d0713191b2887814594c56a4f169e4538d/packages/data/src/components/use-dispatch/use-dispatch-with-map.js#L39

Reply:
ya that's what I expected too.  But both the e2e test and manual testing were failing for that behaviour (when using the keyboard shortcut for that modal when it was open, `isModalActive` was always false - as if it was closed).  It might be that this is exposing an issue somewhere else?

Reply:
Yeah I found the reason why.

https://github.com/WordPress/gutenberg/blob/13e58516cc34dfd3fabc569ef2177c332091964e/packages/components/src/keyboard-shortcuts/index.js#L25

That component binds the callbacks on mount and never updates them.

I guess we'll have to keep the ref then. But I think we should reference the ref instead of the variable everywhere for consistency. We should also use `useLayoutEffect` instead of `useEffect` to avoid action dispatchers called between renders and effects having an outdated callback.

Yeah it's getting pretty late here too. Good night!